src/server/scene/basic_surface.cpp

```diff
@@ -377,22 +377,42 @@ bool ms::BasicSurface::input_area_contains(geom::Point const& point) const
         auto const local_point = as_point(point - content_top_left(*state));
         for (auto const& rectangle : state->custom_input_rectangles)
         {
+            // Note that this size is already rotated according to the orientation, so the
+            // width and height do not need to be swapped.
+            auto const& surface_size = state->surface_rect.size;
+
             // When a surface is rendered with an orientation, it is unrotated so that it appears
-            // upright. The input region that the surface provides is also rotated according to the
-            // same rules. Hence, we need to swap the width and height of the input region if the
-            // surface is rotated 90 or 270 degrees.
-            if (state->orientation == mir_orientation_right || state->orientation == mir_orientation_left)
+            // upright. The input region that the surface provides is given in surface-local coordinates
+            // meaning that it will need to be rotated to match up with the surface.
+            switch (state->orientation)
+            {
+            case mir_orientation_left:
             {
                 auto rotated_rect = rectangle;
                 rotated_rect.top_left.x = geom::X(rectangle.top_left.y.as_value());
                 rotated_rect.top_left.y = geom::Y(rectangle.top_left.x.as_value());
                 rotated_rect.size.width = geom::Width(rectangle.size.height.as_value());
                 rotated_rect.size.height = geom::Height(rectangle.size.width.as_value());
-                if (rotated_rect.contains(local_point))
-                    return true;
+                return rotated_rect.contains(local_point);
+            }
+            case mir_orientation_right:
+            {
+                auto rotated_rect = rectangle;
+                rotated_rect.top_left.x = geom::X(surface_size.width.as_value() - rectangle.top_left.y.as_value() - rectangle.size.height.as_value());
+                rotated_rect.top_left.y = geom::Y(surface_size.height.as_value() - rectangle.top_left.x.as_value() - rectangle.size.width.as_value());
+                rotated_rect.size.width = geom::Width(rectangle.size.height.as_value());
+                rotated_rect.size.height = geom::Height(rectangle.size.width.as_value());
+                return rotated_rect.contains(local_point);
+            }
+            case mir_orientation_inverted:
+            {
+                auto rotated_rect = rectangle;
+                rotated_rect.top_left.y = geom::Y(surface_size.height.as_value() - rectangle.top_left.y.as_value() - rectangle.size.height.as_value());
+                return rotated_rect.contains(local_point);
+            }
```

Comment:
This looks wrong:

Suppose the input rectangle is a LHS border, then inverted is a 180 degree rotation and the transformed rectangle should be a RHS border.

This transform keeps the rectangle (x is unchanged) on the left and moves it below the window (y goes from 0 to `surface_size.height`)